include/alpaka/mem/buf/Traits.hpp

```diff
@@ -191,4 +191,64 @@ namespace alpaka
 
         ALPAKA_UNREACHABLE(allocBuf<TElem, TIdx>(host, extent));
     }
+
+    namespace detail
+    {
+        template<typename DevDst, typename DevSrc>
+        auto canZeroCopy(const DevDst& devDst, const DevSrc& devSrc) -> bool
+        {
+            if constexpr(std::is_same_v<DevDst, DevSrc>)
+                if(devSrc == devDst)
+                    return true;
+            return false;
+        }
+    } // namespace detail
+
+    //! Makes the content of the source view available on the device associated with the destination queue. If the
+    //! destination shares the same memory space as the source view, no copy is performed and the destination view is
+    //! updated to share the same buffer as the source view. Otherwise, a memcpy is performed.
+    template<typename TQueue, typename TViewDst, typename TViewSrc>
+    ALPAKA_FN_HOST void zero_memcpy(TQueue& dstQueue, TViewDst& viewDstOpt, TViewSrc const& viewSrc)
```

Comment:
This version is used when the user already has a destination buffer. Typically, the case when transferring back from the device to the host buffer.

Reply:
From our offline discussion yesterday: IMO `TQueue& dstQueue` must be the source queue because you should not provide the control back to the caller before the memory operations on the source device are finished.

Reply:
I'm thinking of two possible use cases:
  - "transfer" data from _host_ to _device_
  - "transfer" data from _device_ to _host_

where _device_ could be any device for which we do not want to make the copy, but take the zero-copy path; it could be from a CPU backend, or a GPU backend with unified memory, _etc_.

In both cases I would expect to use a _queue_ associated to the _device_, not to the _host_, same as for the regular `memcpy` function.

So, maybe this should simply be `TQueue queue` ?

Reply:
@fwyzard I think it must always be the source queue because there zero copy operation is instantly finished, so it takes no time but is not allowed use `viewDstOpt` before all operations on `viewSrc` are finished. To guard this case you need to wait for the source queue. In cases we would introduce a `zero_memcpy_async` we would need to return a event which provide the information when we can use `viewSrc`. 
Even for `host to device` we need the source queue (host queue) because the user can run a host accelerated kernel with `viewSrc`.

Reply:
Mhh I am a little bit confused the function is called `zero_memcpy` but is performing a copy in case the memory is not available on the destination device. So it is **NOT** zero-memcopy, it is some kind of an optimized (intelligent) copy.

In that case I think we need the source and destination queue to gard when the operation starts in case it is not `zero copiable` and to secure that follow-up operations to the destination not start before the possible copy is finished.

IMO the function should not be called `zero_memcpy` or I have another definition of zero memcopy in mind.

Reply:
> In both cases I would expect to use a queue associated to the device, not to the host, same as for the regular memcpy function.
> So, maybe this should simply be TQueue queue ?

Exactly, so I renamed it yesterday like so, but GitHub is only showing `Outdated` instead of the actual state of the PR.

> it is not allowed to use viewDstOpt before all operations on viewSrc are finished. To guard this case you need to wait for the source queue.

So how do we guard against this in the standard `alpaka::memcpy(queue, dstView, srcView)`? Isn't it the same problem here? So I wondered what the meaning of this queue actually is, and whether it needs to have any relation to the views. Looking into the implementation, there is actually no requirement on the queue related to the views. For CUDA, the associated devices of the views determine whether the memcpy kind is host-2-device, device-2-device, etc., but I could apparently use a queue to a device A and then memcpy between two views on different devices B and C. This may be a design bug, but it leads me to believe that the queue's meaning is to allow the user to sequence operations with respect to other operations. And if the user gets that wrong, then that's the users fault.

> Even for host to device we need the source queue (host queue) because the user can run a host-accelerated kernel with viewSrc.

Could you point me to an example where we actually have a queue to the host device as well (in addition to the accelerator device) and how that would be used? This is the first time I hear of this use.

> Mhh I am a little bit confused the function is called zero_memcpy but is performing a copy in case the memory is not available on the destination device.

Right, which is why I chose the provocative placeholder name `zero_memcpy`. We still need to have that discussion on how to name this functionality.

I was tempted yesterday to define zero-copy as: `zero_memcpy(queue, dev, srcView) -> optional<DstView>` or even omit the `queue` parameter altogether, because there is no real work necessary in case the view is also accessible on the destination device. However, this is not the API I would want to have as a user and I would immediately need to implement:
```c++
if (auto dst = zero_memcpy(queue, dev, srcView))
    return *dst;
else {
    auto dst = allocBuf(...);
    memcpy(queue, dst, src);
}
```
I think @j-stephan suggested in the past to use `make_available` and @fwyzard suggested to use `convey` or `smart_memcpy` in #28. 

> In that case I think we need the source and destination queue to gard when the operation starts in case it is not zero copiable and to secure that follow-up operations to the destination not start before the possible copy is finished.

Again, this is the users responsibilty to use the queues correctly. If you create two queues and use two views on both of the queues, you can also create a race with just the normal `alpaka::memcpy`.

Reply:
> Could you point me to an example where we actually have a queue to the host device as well (in addition to the accelerator device) and how that would be used? This is the first time I hear of this use.

I would say we do not have such an example because we do not use accelerators colocated in one application. 
As I know we are very "dirty" in all our examples and run host code without a use of an accelerator because we run the host code sequentially. 
Even in PIConGPU we use native OpenMP where normally an accelerator would be the better way to run parallel host code to re-use all device functions we implemented. The problem with PIConGPU is that we use `cupla` where we can not use two accelerators at the same time.
To be fair but this is one reason why we made alpaka but we never created an example :-(
Our tests are using multiple accelerators in one test-app but not cooperative as I know :-(

Reply:
> As I know we are very "dirty" in all our examples and run host code without a use of an accelerator because we run the host code sequentially.

I don't think this is dirty. Alpaka should just coexist with ordinary C++. If I use a `std::for_each(std::execution::par, ...)` or a `std::jthread` next to an alpaka kernel call, this should be allowed (given that I also get all my synchronization right).

> Our tests are using multiple accelerators in one test-app but not cooperative as I know :-(

Yeah, I think we should have a test/example for that, see #1909.

Reply:
Yes it should co-exists with "normal" C++. Nver the less all our examples assume we do not need paralism on the host side and we simply use a sequential for loop e.g. to evaluate results instead of a kernel that can do it in parallel.